bfvmm/include/hve/arch/intel_x64/vmcs.h

```diff
@@ -205,75 +186,14 @@ class EXPORT_HVE vmcs
     /// @return returns true if the VMCS is configured properly, false
     ///     otherwise
     ///
-    bool check() const noexcept;
-
-    /// Save State
-    ///
-    /// Returns the VMCS's save state. This is state that is above and beyond
-    /// what the VMCS stores, includng the CPU's registers, vcpuid and
-    /// exit handler pointer.
-    ///
-    /// @expects none
-    /// @ensures none
-    ///
-    /// @return returns the VMCS's save state.
-    ///
-    save_state_t *save_state() const
-    { return m_save_state.get(); }
-
-    /// MSR bitmap
-    ///
-    /// Returns the VMCS's msr bitmap
-    ///
-    /// @expects none
-    /// @ensures none
-    ///
-    /// @return returns the VMCS's msr bitmap
-    ///
-    gsl::not_null<uint8_t *> msr_bitmap() const
-    { return m_msr_bitmap.get(); }
-
-    /// IO bitmap a
-    ///
-    /// Returns the VMCS's io bitmap a
-    ///
-    /// @expects none
-    /// @ensures none
-    ///
-    /// @return returns the VMCS's io bitmap a
-    ///
-    gsl::not_null<uint8_t *> io_bitmap_a() const
-    { return m_io_bitmap_a.get(); }
-
-    /// IO bitmap b
-    ///
-    /// Returns the VMCS's io bitmap b
-    ///
-    /// @expects none
-    /// @ensures none
-    ///
-    /// @return returns the VMCS's io bitmap b
-    ///
-    gsl::not_null<uint8_t *> io_bitmap_b() const
-    { return m_io_bitmap_b.get(); }
+    VIRTUAL bool check() const noexcept;
 
 private:
 
-    page_ptr<save_state_t> m_save_state;
+    vcpu *m_vcpu;
```

Comment:
👎  Why did this circular relationship make it back into the code again?

Reply:
Both the local and global state structures for the vCPU are stored in the vCPU as these are not VMCS objects. The launch() function needs a pointer to the vCPU to get access to some of this information (register data really). Although it is circular, it is technical "weak" which is fine. We could remove all weak references, but it would really mess up the function calls as you would have to pass every object to every function for all classes, which doesn't make any sense. 

Most of the circular references were removed, and we pass the vCPU around where it makes sense. In this case, every function needs the vCPU and since its a "weak" reference, it is guideline compliant, safe, and free of any construction ordering issues. IMO, this provides a nice balance/compromise, and later on, we could always continue to remove more of these over time as needed.

Reply:
My opinion is that all of these weak vcpu references are causing a real headache in terms of separating concerns. If a vcpu needs to know how to control a vmcs, and a vmcs needs to know how to control a vcpu, then they might as well be the same object.

Reply:
Not really for a couple of reasons:
- The vCPU provides the public API that an object needs. In C, this is the same thing as including a header in a circular manner which is done on a regular basis for the same reason. The circular reference is built into Intel's spec and will be there one way or another. The IO instruction code is a great example of this. There is no way to prevent the circular reference as Intel designed it that way. 
- Qt is a great example of why a "weak" reference is an important tool. It breaks the code up into logical components that are easier to test on their own instead of all in one. Yes, putting all of the code into the vCPU would remove the weak references, but it would create a giant blob of code that would be hard to test. Over time, our goal should be to move code from the vCPU into their own classes, even if a weak reference is needed to do it, for the sole purpose of ensuring that a class only has a single purpose/job, which is a common design pattern of C++. 

As for separating, there is no need to do so. A vCPU cannot function without the VMCS and vice versa, and there really isn't a need in the code to find a way to separate them. With respect to APIs and testing, nothing is affected negatively so IMO, there really is no headache here as there is really nothing to do/fix. For a lack of better words, I fail to see the problem it creates. Maybe a better way to say it is, there is a lot for us to fix/work on, this is very low on the priority list for me to fix since it is not getting in our way, and attempting to fix this now clutters our interfaces and pulls us from other more important things.

Reply:
> * but it would create a giant blob of code that would be hard to test

Correct, this is what the vCPU already is. I just disagree that these week references are the best way to solve this problem

> * Over time, our goal should be to move code from the vCPU into their own classes

Agreed, and thats the most important part. As long as thats the end goal, lets leave this weak reference for now